File: pkg-py/tests/test_conformance.py
"""Run the shared cross-language conformance corpus against the Python package."""

import json
from importlib.resources import files

import pytest

import biobouncer


def _load_cases():
    root = files("biobouncer") / "_data" / "corpus"
    cases = []
    for entry in sorted(root.iterdir(), key=lambda p: p.name):
        if not entry.name.endswith(".jsonl"):
            continue
        for line in entry.read_text(encoding="utf-8").splitlines():
            line = line.strip()
            if line:
                cases.append(json.loads(line))
    return cases


CASES = _load_cases()


@pytest.fixture(autouse=True)
def _isolate_cache(tmp_path, monkeypatch):
    # Empty cache dir, so cache cases resolve to the bundled sample snapshot.
    monkeypatch.setenv("BIOBOUNCER_CACHE_DIR", str(tmp_path))


def test_corpus_is_not_empty():
    assert CASES, "no conformance cases were found in the vendored corpus"


@pytest.mark.parametrize(
    "case", CASES, ids=[f"{c['source_db']}-{c['how']}-{c['input']}" for c in CASES]
)
def test_conformance(case):
    expect = case["expect"]
    result = biobouncer.check_id(
        case["input"],
        source_db=case["source_db"],
        how=case["how"],
        species=case.get("species"),
        version=case.get("version"),
    )[0]
    assert result.valid == expect["valid"]
    assert result.normalized == expect.get("normalized")
    assert result.suggestion == expect.get("suggestion")

File: pkg-r/tests/testthat/test-conformance.R
test_that("cross-language conformance corpus passes", {
  skip_if_not_installed("jsonlite")
  # Isolate the cache dir so cache cases resolve to the bundled sample snapshot.
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())

  dir <- system.file("extdata", "corpus", package = "biogate")
  files <- sort(list.files(dir, pattern = "\\.jsonl$", full.names = TRUE))
  expect_gt(length(files), 0)

  n_cases <- 0
  for (f in files) {
    lines <- readLines(f, warn = FALSE)
    lines <- lines[nzchar(trimws(lines))]
    for (line in lines) {
      case <- jsonlite::fromJSON(line, simplifyVector = FALSE)
      res <- check_id(
        case$input,
        source_db = case$source_db,
        how = case$how,
        version = case$version
      )
      exp <- case$expect
      exp_norm <- if (is.null(exp$normalized)) NA_character_ else exp$normalized
      exp_sugg <- if (is.null(exp$suggestion)) NA_character_ else exp$suggestion
      info <- sprintf("%s %s: %s", case$source_db, case$how, case$input)
      expect_identical(res$valid, exp$valid, info = info)
      expect_identical(res$normalized, exp_norm, info = info)
      expect_identical(res$suggestion, exp_sugg, info = info)
      n_cases <- n_cases + 1
    }
  }
  expect_gt(n_cases, 0)
})
